### PR TITLE
[SPARK-18471][MLLIB] In LBFGS, avoid sending huge vectors of 0

### PR DESCRIPTION
**Pull request by AnthonyTruchet**
## What changes were proposed in this pull request?

CostFun used to send a dense vector of zeroes as a closure in a
treeAggregate call. To avoid that, we replace treeAggregate by
mapPartition + treeReduce, creating a zero vector inside the mapPartition
block in-place.

## How was this patch tested?

Unit test for module mllib run locally for correctness.

As for performance we run an heavy optimization on our production data (50 iterations on 128 MB weight vectors) and have seen significant decrease in terms both of runtime and container being killed by lack of off-heap memory.

### CODE REVIEW

**Comment by AnthonyTruchet** [comment]
@srowen Here is at last the real PR for SPARK-18471. 

Sorry for the noise due to GitHub fiddling...

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala [hunk 1]
```diff
@@ -241,16 +241,27 @@ object LBFGS extends Logging {
       val bcW = data.context.broadcast(w)
       val localGradient = gradient
 
-      val (gradientSum, lossSum) = data.treeAggregate((Vectors.zeros(n), 0.0))(
-          seqOp = (c, v) => (c, v) match { case ((grad, loss), (label, features)) =>
-            val l = localGradient.compute(
-              features, label, bcW.value, grad)
+      /** Given (current accumulated gradient, current loss) and (label, features)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala]: Nit: just use `//` for two lines of comment. Really `/**` (vs `/*`) is for javadoc.

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala [hunk 2]
```diff
@@ -241,16 +241,27 @@ object LBFGS extends Logging {
...
+
+      // Adds two (gradient, loss) tuples
+      val combOp = (c1: (Vector, Double), c2: (Vector, Double)) =>
+        (c1, c2) match { case ((grad1, loss1), (grad2, loss2)) =>
             axpy(1.0, grad2, grad1)
             (grad1, loss1 + loss2)
-          })
+       }
+
+      val (gradientSum, lossSum) = data.mapPartitions { it => {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala]: Nit: the second brace and its match are redundant

**Review by srowen — CHANGES_REQUESTED** [review]
I think it's still worth looking into making a similar change elsewhere that seqOp and combOp are used this way, and which might also accidentally take something into the closure. But (modulo minor changes) I think this is an OK change.

**Comment by srowen** [comment]
Ping @AnthonyTruchet

**Comment by AnthonyTruchet** [comment]
Hello @srowen , sorry not to have updated this lately, I've been taken by other emergencies at work. I'll update this on Monday. Actually I'll submit a variant of treeAggregate in core that we will be able to use for other similar use case in ML(lib). 

I appreciate your care about those patches and will attempt to reach a reasonable reactivity too :-)

**Comment by AnthonyTruchet** [comment]
Once more (last time hopefully) I mistakenly fiddled with PR. Closing this one and replace it with #16037.
Code style review above taken into account in new PR.

**Pull request closed without merging**